### PR TITLE
Component review and improvements

### PR DESCRIPTION
**Pull request by FabienDeshayes**
Following the review of the accordion by the team, we added / removed some bits to match our functional and non-functional requirements.

## Description
The review contains the following changes:

* Remove CSS provided in the component to be the minimum
* Add and document custom header and custom body use cases
* Fix issue where getOpenItems() would returns too many items because the `active` property of the state was not removing old items
* Fix and improve examples
* Sort getOpenitems() result

## Motivation and Context
To follow our functional and non-functional requirements, mainly the following:

* A component should have accompanying examples
* By default a component must produce a rendered output: it must not be faceless when a specific skin hasn't been applied to it yet, i.e. must have minimal styling to visually represent itself (vanilla theme)
* When extension points are necessary, a component must provide documentation on how to such extensions can be implemented, i.e. plugins, etc...

## How Was This Tested?
We run the examples using the webpack dev server, and run the Jest test suite.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My change follows the style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the [contribution guidelines](../CONTRIBUTING.md)
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by FabienDeshayes** [comment]
So the work was 95% done from @DimitarChristoff, we just fixed / added of couple of extras like example / documentation.

The main point I'd like to discuss is around CSS. We've made a conscious decision to have minimal styles in the component itself, but there are two outstanding questions:

* Should we inline the component styles that we want to "force", i.e. prevent the user from overriding (unless they really need to)? E.g. setting `height` / `max-height` on the accordion item
* How much of cosmetics / vanilla theme style do we want to provide? At the moment we've got a `background-color` and `color` for the headers, plus some `padding` to make it looking reasonably good by default. It also has a simple `transition` when the accordion opens / closes. But is that already too much, or do we consider these to be part of the vanilla theme?

My opinion is we should inline what we want to set in stone and probably keep the vanilla theme as it is (maybe with some cleaning of styles for nesting / namespacing).

**Comment by grancalavera** [comment]
> Should we inline the component styles that we want to "force"

I would say yes

> Should we inline the component styles that we want to "force"

As much as you want but keeping it to the minimum necessary, as long as they can be overridden and the selectors needed to do so are documented.

File: src/css/index.css [hunk 1]
```diff
@@ -1,129 +1,34 @@
-@import url(http://fonts.googleapis.com/css?family=Lato:400,700);
+.zb-accordion {
```
Comment (grancalavera) [inline src/css/index.css]: I think we need to make selectors that refer to components distinct from other possible selectors, for instance selectors that may refer to some generic property `zb-selected`. I'm playing around with a verbose option `zb-component-accordion`. But  something sorter should work as well `zbc-[name]` for components... something along those lines.

File: examples/CustomRendererExample.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import                           './example.css'
```
Comment (grancalavera) [inline examples/CustomRendererExample.js]: Remove extra spaces.

File: examples/index.js [hunk 3]
```diff
@@ -2,9 +2,15 @@ import '../src/css/index.css'
 import React      from 'react'
 import ReactDOM   from 'react-dom'
 import Example    from './Example'
+import Custom     from './CustomRendererExample'
```
Comment (grancalavera) [inline examples/index.js]: Remove extra spaces

File: src/css/index.css (hunk 1 above)
Comment (DimitarChristoff) [inline src/css/index.css]: probably a good idea. accordion is way too generic.

Comment (FabienDeshayes) [inline src/css/index.css]: I've gone with `zbc-accordion` and `zbc-accordion-title`, `zbc-accordion-item`.
The accordion itself has the `zb-component` class name as well.
Let's discuss this to agree on the naming.

File: examples/CustomRendererExample.js (hunk 2 above)
Comment (FabienDeshayes) [inline examples/CustomRendererExample.js]: I actually don't mind the spaces, I guess it's just personal preference. Worth removing to be consistent with our other repos?

**Comment by mstade** [comment]
> - Should we inline the component styles that we want to "force", i.e. prevent the user from overriding (unless they really need to)? E.g. setting height / max-height on the accordion item

Yes. Any styles that are **required** for **functional** reasons should not be overridable, or at least very difficult to override. If there are better ideas for how to do this other than increased specificity, I'm all ears.

The imperative question here is: how do you determine if a style is *functionally required*? I bet that question can be difficult to answer at times, and trivial at others. Moreover, I'd guess that whatever this set of styles are, it's probably going to be *very* small.

> - How much of cosmetics / vanilla theme style do we want to provide?

None. Other than styles required for functional reasons should not be included, even if that makes the component ugly as sin. Styles applied using selectors should obviously be documented well since they make up part of the API, and that documentation can very well be thought of as a theme. (Even if that theme only styles this component; the end result of that is that concatenating all component theme styles would then make up a whole theme composition.) But if you render this component, and have not loaded a theme, the only styles applied should be the functional styles, those that should never, ever be overridden, even if it does mean the thing is ugly as sin.

At least, that's how I've interpreted the spec and discussions we've had so far.

An analogy I've been using in my head is that applying the component without a theme is like rendering a web page without linking any stylesheets. It'll be ugly as sin probably, because it will render whatever the browser defaults are. The functional styles are like browser defaults (except they can't be overridden, so I guess the analogy breaks down a bit) and themes are like linking css files. Maybe that makes more sense in my head than it does written down...

File: src/css/index.css (hunk 1 above)
Comment (mstade) [inline src/css/index.css]: My take on this is that it probably doesn't matter so much, so long as it's consistent. The terser we are, the more chance of collisions with random css, but perhaps we should try and get some practical experience around it to see how much of a risk that actually is – i.e. there's probably less risk in using a prefix like `.zambezi-` than `.zb-`, but if the risk is negligible anyway, does it really matter? Can we simulate scoping somehow to reduce the risk further in any event, yet not adversely affect performance?

**Comment by FabienDeshayes** [comment]
@Poetro @mstade I think we should continue this discussion on our own GitHub instance, otherwise we'll loose commit / information

**Pull request closed without merging**